src/schema/FroidSchema.ts

```diff
@@ -0,0 +1,839 @@
+   *
+   * Contract @tag directives are NOt applied when generating non-native scalar
+   * return types in the Froid subgraph. Contract @tag directives are merged
+   * during supergraph composition so Froid subgraph can rely on @tag directives
+   * defined by the owning subgraph(s), UNLESS an enum value is marked @inaccessible,
+   * which is only applicable in Federation 2 schemas.
+   *
+   * @returns {DefinitionNode[]} The non-native scalar definitions needed to be definied in the froid subgraph
+   */
+  private createCustomReturnTypes(): SupportedFroidReturnTypes[] {
+    const froidNodes = Object.values(this.froidObjectTypes);
+    const froidNodeNames = froidNodes.map((obj) => obj.node.name.value);
+
+    // Extract field return values that aren't native scalars (int, string, boolean, etc.)
+    // and isn't a type that is already defined in the froid subgraph
+    const nonNativeScalarDefinitionNames = new Set<string>();
+    froidNodes.forEach((obj) => {
+      obj.selectedFields.forEach((field) => {
+        const fieldReturnType = FroidSchema.extractFieldType(field);
+        if (
+          !scalarNames.includes(fieldReturnType) &&
+          !froidNodeNames.includes(fieldReturnType)
+        ) {
+          nonNativeScalarDefinitionNames.add(fieldReturnType);
+        }
+      });
+    });
+
+    // De-dupe non-native scalar return types. Any definitions of scalars and enums
+    // will work since they can be guaranteed to be consistent across subgraphs
+    const nonNativeScalarFieldTypes = new Map<
+      string,
+      SupportedFroidReturnTypes
+    >();
+    (
+      this.filteredDefinitionNodes.filter((definitionNode) =>
+        typeDefinitionKinds.includes(definitionNode.kind)
+      ) as SupportedFroidReturnTypes[]
+    ).filter((nonNativeScalarType) => {
+      const returnTypeName = nonNativeScalarType.name.value;
+      // Get only types that are returned in froid schema
+      if (
+        nonNativeScalarDefinitionNames.has(returnTypeName) &&
+        !nonNativeScalarFieldTypes.has(returnTypeName)
+      ) {
+        if (nonNativeScalarType.kind === Kind.ENUM_TYPE_DEFINITION) {
+          const enumValues = nonNativeScalarType.values?.map((enumValue) => ({
+            ...enumValue,
+            directives: enumValue.directives?.filter(
+              (directive) => directive.name.value === 'inaccessible'
+            ),
+          }));
+          nonNativeScalarFieldTypes.set(returnTypeName, {
+            ...nonNativeScalarType,
+            values: enumValues,
+            directives: [],
+            description: undefined,
+          } as EnumTypeDefinitionNode);
+        } else if (nonNativeScalarType.kind === Kind.SCALAR_TYPE_DEFINITION) {
+          nonNativeScalarFieldTypes.set(returnTypeName, {
+            ...nonNativeScalarType,
+            description: undefined,
+            directives: [],
+          } as ScalarTypeDefinitionNode);
+        }
+      }
+    });
```

Comment:
does not look it filter anything  -> it is more an iteration ?

```suggestion
    (
      this.filteredDefinitionNodes.filter((definitionNode) =>
        typeDefinitionKinds.includes(definitionNode.kind)
      ) as SupportedFroidReturnTypes[]
    ).forEach((nonNativeScalarType) => {
      const returnTypeName = nonNativeScalarType.name.value;

      // Get only types that are returned in froid schema
      const isTypeFromFroidSchema = nonNativeScalarDefinitionNames.has(returnTypeName) &&
        !nonNativeScalarFieldTypes.has(returnTypeName);
      if (! isTypeFromFroidSchema) {
        return;
      }
            
      if (nonNativeScalarType.kind === Kind.ENUM_TYPE_DEFINITION) {
        const enumValues = nonNativeScalarType.values?.map((enumValue) => ({
          ...enumValue,
          directives: enumValue.directives?.filter(
            (directive) => directive.name.value === 'inaccessible'
          ),
        }));
        nonNativeScalarFieldTypes.set(returnTypeName, {
          ...nonNativeScalarType,
          values: enumValues,
          directives: [],
          description: undefined,
        } as EnumTypeDefinitionNode);
        return;
      }
      
      if (nonNativeScalarType.kind === Kind.SCALAR_TYPE_DEFINITION) {
        nonNativeScalarFieldTypes.set(returnTypeName, {
          ...nonNativeScalarType,
          description: undefined,
          directives: [],
        } as ScalarTypeDefinitionNode);
        return;
      }
    });
```

Reply:
That's a fair observation. For what it's worth, I pulled in existing code as methods to this class to give me the freedom to change anything I needed to change. This is one of those functions. I'll take a look at revising this method.